Improve namespace packaging guide (#1379)
pypa/packaging.python.org: source/guides/packaging-namespace-packages.rst

Creating a namespace package
============================

There are currently two different approaches to creating namespace packages,
from which the latter is discouraged:

#. Use `native namespace packages`_. This type of namespace package is defined
   in :pep:`420` and is available in Python 3.3 and later. This is recommended if
   packages in your namespace only ever need to support Python 3 and
   installation via ``pip``.
#. Use `legacy namespace packages`_. This comprises `pkgutil-style namespace packages`_
   and `pkg_resources-style namespace packages`_.

Native namespace packages
-------------------------

Python 3.3 added **implicit** namespace packages from :pep:`420`. All that is
required to create a native namespace package is that you just omit
:file:`__init__.py` from the namespace package directory. An example file
structure (following :ref:`src-layout <setuptools:src-layout>`):

.. code-block:: text

    mynamespace-subpackage-a/
        pyproject.toml # AND/OR setup.py, setup.cfg

>         src/
>             mynamespace/ # namespace package
>             # No __init__.py here.
>                 subpackage_a/
>                     # Sub-packages have an __init__.py.

Reviewer: I would align indentation like this:

```suggestion
        src/
            mynamespace/ # namespace package
                # No __init__.py here.
                subpackage_a/
                    # Sub-packages have an __init__.py.
```

>                     # Sub-packages have an __init__.py.

Reviewer: ```suggestion
                    # Regular import packages have an __init__.py.
```

                    __init__.py
                    module.py

It is extremely important that every distribution that uses the namespace
package omits the :file:`__init__.py` or uses a pkgutil-style
:file:`__init__.py`. If any distribution does not, it will cause the namespace
logic to fail and the other sub-packages will not be importable.

The ``src-layout`` directory structure allows automatic discovery of packages

> by most :term:`build backends <Build Backend>`. If however you want to manage

Reviewer: Add link to "[_src layout vs flat layout_](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)"? It does not seem to have a target yet though, so it might need to be added.

Author: Done!

for more information. If however you want to manage exclusions or inclusions of packages
yourself, this is possible to be configured in the top-level :file:`pyproject.toml`:

.. code-block:: toml

    [build-system]
    requires = ["setuptools", "setuptools-scm"]
    build-backend = "setuptools.build_meta"

> 

Reviewer: If the goal of this code block is to show how to configure setuptools to handle namespace packages, then maybe we can skip the `build-system` table. What do you think?

Author: How about the following?

```suggestion
    [build-system]
    ...

```

Reviewer: I tend do write `# ...` or `# [...]` in such cases. But I am fine with anything, even the whole build system table, I don't mind much, it is just nit-picking.

Author: Using my suggestion since that comes to you use in several other places.

    [tool.setuptools.packages.find]
    where = ["src/"]
    include = ["mynamespace.subpackage_a"]

    [project]
    name = "mynamespace-subpackage-a"
    ...

The same can be accomplished with a :file:`setup.cfg`:

.. code-block:: ini

>    [options]

Reviewer: ```suggestion
    [options]
```

.. code-block:: text

    mynamespace-subpackage-a/

>         setup.py # AND/OR pyproject.toml, setup.cfg

Reviewer: I would for sure not put `setup.py` in first place here. My personal preference is to only put standardized things in `pyproject.toml` so I would put `setup.cfg` in first place, but I guess the preferred way in the community is to put as much as possible in `pyproject.toml` so that is probably what should be done here as well.

        name='mynamespace-subpackage-a',
        ...
        packages=find_namespace_packages(include=['mynamespace.*'])
        # or list a single package explicitly:
        # packages=['mynamespace.subpackage_a'],
    )

The same can be accomplished by replacing the :file:`setup.py` in the

> namespace packages' parent directory with a :file:`pyproject.toml`,

Reviewer: ```suggestion
project's source tree root directory with a :file:`pyproject.toml`,
```

    setup(
        name='mynamespace-subpackage-a',
        ...
        packages=find_namespace_packages(where='src/', include=['mynamespace.subpackage_a']),

>         package_dir={"": "src"},

Reviewer: Fixing quotes for consistency of this clock of code.
Maybe we should have some formatting rules for the whole guide?

```suggestion
        package_dir={'': 'src'},
```

Author: Yes, we might dig out a hook somewhere.

Reviewer: Something that formats Python code embedded inside reStructuredText? No idea if that exists, but maybe it does exist, I have no idea.

    only support Python 3 and pkgutil-style namespace packages in the
    distributions that need to support Python 2 and 3.

Legacy namespace packages
-------------------------

> These to methods, that were used to create namespace packages prior to :pep:`420`,

Reviewer: ```suggestion
These two methods, that were used to create namespace packages prior to :pep:`420`,
```

.. code-block:: text

    mynamespace-subpackage-a/
        src/
            pyproject.toml # AND/OR setup.cfg, setup.py
            mynamespace/
                __init__.py  # Namespace package __init__.py
                subpackage_a/

>                     __init__.py  # Sub-package __init__.py

Reviewer: ```suggestion
                    __init__.py  # Regular package __init__.py
```

Author: I find this a bit confusing, since the namespace package _is_ a sub-package and the root directory is named accordingly?

Probably 
```suggestion
                    __init__.py  # Regular sub-package __init__.py
```

is more like it ;-)

Reviewer: My reasoning for this is that from the point of view of the Python interpreter there are two kinds of importable packages: the regular ones without qualifier and the "namespace" packages.

Whether they are sub-packages or not does not matter.

For example:

```none
.
├── namespace
│   └── regular
│       └── __init__.py
└── regular
    ├── __init__.py
    └── namespace
```

```pycon
>>> import namespace
>>> namespace
<module 'namespace' (<_frozen_importlib_external.NamespaceLoader object at 0x7fb3d531f090>)>
>>> import namespace.regular
>>> namespace.regular
<module 'namespace.regular' from '/tmp/tmp.wRR0QJjvej/namespace/regular/__init__.py'>
>>> import regular
>>> regular
<module 'regular' from '/tmp/tmp.wRR0QJjvej/regular/__init__.py'>
>>> import regular.namespace
>>> regular.namespace
<module 'regular.namespace' (<_frozen_importlib_external.NamespaceLoader object at 0x7fb3d4b6af90>)>
```

I know that this is valid for PEP420 native namespace packages, and this example is not PEP420, but from my point of view we should try to stick to this terminology.

I guess I am fine with "regular sub-package".

Author: Let's do it like you suggest, then. I think I'm the one who's nit-picking.

.. code-block:: text

    mynamespace-subpackage-a/
        src/
            pyproject.toml # AND/OR setup.cfg, setup.py
            mynamespace/
                __init__.py  # Namespace package __init__.py
                subpackage_a/

>                     __init__.py  # Sub-package __init__.py

Reviewer: ```suggestion
                    __init__.py  # Regular package __init__.py
```
